### ISSUE TITLE
Shorten the waiting time of GitHub CI

### ISSUE THREAD

**Issue by mytting** [deepflowys/deepflow#1101]
### Feature request

Shorten the waiting time of GitHub CI

**Comment by mytting** [comment]
@aktech 
Hello, We have fully launched cirun.io and abandoned the original runner, and found some problems today.
We think the current one and a half minute response of github CI is a bit slow, is there a way to speed it up? For example, do some preset software in ami and so on? In addition, we have a problem that the runner requested by CI is taken away by other CI. Is there a good solution?

**Comment by aktech** [comment]
> We think the current one and a half minute response of github CI is a bit slow, is there a way to speed it up? For example, do some preset software in ami and so on?

Hey @mytting yes you can create custom AMI with some of the software already installed like say docker, etc. That would speed up your overall CI time. The provision time wouldn't be affected much as it's mainly just calling AWS's API spinning up a VM and installing Git Actions, installation doesn't take much time, like less than 15-20 seconds. Most of the time is spent getting a VM from AWS. I can try to take a look if there are any bottlenecks which can be improved.

> In addition, we have a problem that the runner requested by CI is taken away by other CI. Is there a good solution?

What do you mean by other CI? Do you mean other jobs? Runners are are picked up by GitHub Action workflows by runner labels, which is controlled by:

```yml
runs-on: cirun-aws-amd64-32c
``` 

If you want them to be unique I can look on implementing spinning up runners by run_id, then you could do something like:

```yml
# Not implemented yet
runs-on: cirun-aws-amd64-32c--${{ github.run_id }}"
```

will that help?

**Comment by mytting** [comment]
```cirun-aws-amd64-32c--${{ github.run_id }}"```
Yes, I want this effect. Is there anything you need to do?

@aktech

**Comment by mytting** [comment]
When will that be possible?

**Comment by aktech** [comment]
> Yes, I want this effect. Is there anything you need to do?

Yes I need to implement it. You should have it within a few days (maximum: a week). I'll implement it at share the documentation link here.

**Comment by mytting** [comment]
> > Yes, I want this effect. Is there anything you need to do?
> 
> Yes I need to implement it. You should have it within a few days (maximum: a week). I'll implement it at share the documentation link here.

ok, wait for the good news. Does github action support this syntax?

**Comment by mytting** [comment]
Another restriction is that the runner label must begin with cirun, as if it is not mentioned in the documentation
@aktech

**Comment by mytting** [comment]
> > > Yes, I want this effect. Is there anything you need to do?
> > 
> > 
> > Yes I need to implement it. You should have it within a few days (maximum: a week). I'll implement it at share the documentation link here.
> 
> ok, wait for the good news. Does github action support this syntax?

Oh, I just tried it. github supports this syntax

**Comment by aktech** [comment]
> Another restriction is that the runner label must begin with cirun, as if it is not mentioned in the documentation

Thanks for pointing that out, I'll update that in the documentation, apologies for the inconvenience.
Yes, that's important because its make my life easier to filter webhook events, where runner needs to be created, otherwise it would have been tricky.

> ok, wait for the good news. Does github action support this syntax?
> Oh, I just tried it. github supports this syntax

Yep, I tried it as well. You would hear from me soon. :)

**Comment by mytting** [comment]
It seems that the spot instance defining multiple regions and multiple specifications in the.cirun file does not work, and the spot request often appears open, 'no Spot capacity available', at which point cirun considers the creation successful.
@aktech

**Comment by aktech** [comment]
Yes, that's an outstanding bug. It will be fixed in the next release.

**Comment by mytting** [comment]
Does cirun support google cloud? aws spot instances are billed by the hour, one hour minimum, our CI usually runs for about 10 minutes, I understand that gcp is billed by the minute,
@aktech

**Comment by aktech** [comment]
> Does cirun support google cloud? 

Yes, it does.

> aws spot instances are billed by the hour, one hour minimum,

Are you sure? to me it seems like you're charged for seconds used: https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/billing-for-interrupted-spot-instances.html

**Comment by mytting** [comment]
> > cirun 支持谷歌云吗？
> 
> 是的，它确实。
> 
> > aws Spot 实例按小时计费，最少一小时，
> 
> 你确定吗？对我来说，您似乎需要为使用秒数付费：[https ://docs.aws.amazon.com/AWSEC2/latest/UserGuide/billing-for-interrupted-spot-instances.html](https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/billing-for-interrupted-spot-instances.html)

I'll check with the sales staff. I think the bill is by the hour

**Comment by aktech** [comment]
> I'll check with the sales staff. I think the bill is by the hour

Quite strange, let me know if you hear from them.

**Comment by mytting** [comment]
> Quite strange, let me know if you hear from them.

Just confirmed that the bill is by the second, was misled by some pages of aws

**Comment by mytting** [comment]
@aktech Hi ,has there been any progress recently？

**Comment by aktech** [comment]
Hey @mytting not yet, I’m travelling at the moment. Expect it by the end of this week.

**Comment by mytting** [comment]
ok

**Comment by aktech** [comment]
Unique runner labels is available now: https://docs.cirun.io/reference/unique-runner-labels

**Comment by mytting** [comment]
Recently, some open source projects came to our aws spot to ask us what to do. We recommended cirun. We will promote cirun recently. cirun has solved many of our pain points. It's great.

**Comment by mytting** [comment]
@aktech Hi 
I have raised an issue to the GitHub documentation, hoping that excellent projects like Cirun can be added to the GitHub documentation, and users will avoid detours
https://github.com/github/docs/issues/21697

In addition, now the job is basically fixed for 90 seconds, and the time may be a bit long. Is there a template for user-data? We can find aws people to see if there is any way to optimize it.

**Comment by aktech** [comment]
Hey @mytting thanks a lot for that, I really appreciate it. There isn't a specific template for the same, but it's fairly simple, which is pulling the github actions runner software and installing it and creating a user and it doesn't take much time really, for example, [here are the logs](https://00f74ba44b48fc7145bb30a8e542b9664b5d889fce-apidata.googleusercontent.com/download/storage/v1/b/cirun/o/provision_logs%2F30845-cirun-deepflowys--deepflow-211e354?jk=AFshE3VmSok-cyZkKC-VhSbsuaQyCMfkc54AwVVB6JMLw6dl1bcoIgN22cP2xDP8eP4j-szmWblHnZJueJd3FxMlmJm2GTqUpxQJC_3iXSXDRXnO-odB-Pfwqiy9nCJ1Ng9vifokLVx7z4KH2Zt-Xv8g-EW6INtfSLfd-T2VTb4SUV2b-5wWzIf9aJECCjb-l5v5T4BsId0ybW-w4taZXkVCa81ras-cFPHlsFBn548Uqul1jIYI3oe_AdIriXu5kRiM2AOc7JI8W5abS-S1hqUC6I1Xza5T6U1fN12ay1RtlKp6i3WvZK5ZylBzB8f3EAm0yrbpw_IOuqxUhdtOfrbuGDAGthwP56htuUIUbYLJTTqxnbLLRpFatPTRxQzBlgRAMb78Its87Wn7flQ2B6GlU9A9zGg34XvG3JMGUZaVMzmFOH7PHS1SVWLMoz2bJXGwlsPtztI76GtU1eOPBXFHpnkH8fA1Iq1B5Ir0caIjtGmSYGcUyV0yrykkCHO0Mfrz1j6jZfJzj4T8IHht5INTfSCM2e-b28cBVPf-3eJutSKKDKOw5SfpABJiVjakxJs8jh6tPmuZPLaaFZ9Eu8Qn5PI7W-pZ1niC4Um2V_7EgI7ZKu2u0wY-Qcs5HuVEsPkJmwyIHo_SY1npcVZUKH5CxebcCxd6Z2XnjL9NLGnIqChD4gKukKZFdAL080viTk1S8F6Je_ar6yR9A9y7QHX9dSNo8xqusFTvxaqhBNr9iwyy8hNZljvoSfE6VH6bpQOZ_rGsMr1LQNfAiekNxHwe9fRTxPR33quwg_thDUsF3K9XflDI0Q36Nr2sVZqE0_0fDtuE3Cprlrd44iav7sjusbt8XAofVmycjq0zgpez7BZ5iBWTNSsiIrc0-0lIctyE04OV31vcjUhEBbPG5_aoCF_dgUSfCyifvJzDVYSISOPRjdyNbixLtS-7DKl0Wnb3hGmaflGeJhedC6xQH8gBNUl6S6Vh7Vo5WhMA1U7Q6nLXF1DWwY8ltScXk5m2VWDswsBRPx0oU3R6JZ-d4qwKARIA4f--WOoS_7P8SkRuTFQKgJ6ZbL6872Ch9dWswemjKMhug2uHw2TM_7i_r3xHgCe3wJik4ujHN8oSiZZALfQPymkV0fSfdaiVaqHP-aMQ_B0KBFlm7gF0D3XHbkWjcFIsLQ6NA49a-LDZWsG0UvPsQNGSbEDxyppveyzJJyOxescdgIGZ4A&isca=1) of a one of the random runners on this repo, it took about 15 seconds for the user data script to run.

My suspicious is on AWS, the time they take to hand over a VM is quite slow. Let me know if you have more questions. I am happy to jump on a call with you and AWS to see where the bottlenecks are.

**Comment by mytting** [comment]
> 

ok ，I have a general understanding, is there a general script? How to pass it to EC2, I can let AWS people debug it

**Comment by mytting** [comment]
> Hey @mytting thanks a lot for that, I really appreciate it. There isn't a specific template for the same, but it's fairly simple, which is pulling the github actions runner software and installing it and creating a user and it doesn't take much time really, for example, [here are the logs](https://00f74ba44b48fc7145bb30a8e542b9664b5d889fce-apidata.googleusercontent.com/download/storage/v1/b/cirun/o/provision_logs%2F30845-cirun-deepflowys--deepflow-211e354?jk=AFshE3VmSok-cyZkKC-VhSbsuaQyCMfkc54AwVVB6JMLw6dl1bcoIgN22cP2xDP8eP4j-szmWblHnZJueJd3FxMlmJm2GTqUpxQJC_3iXSXDRXnO-odB-Pfwqiy9nCJ1Ng9vifokLVx7z4KH2Zt-Xv8g-EW6INtfSLfd-T2VTb4SUV2b-5wWzIf9aJECCjb-l5v5T4BsId0ybW-w4taZXkVCa81ras-cFPHlsFBn548Uqul1jIYI3oe_AdIriXu5kRiM2AOc7JI8W5abS-S1hqUC6I1Xza5T6U1fN12ay1RtlKp6i3WvZK5ZylBzB8f3EAm0yrbpw_IOuqxUhdtOfrbuGDAGthwP56htuUIUbYLJTTqxnbLLRpFatPTRxQzBlgRAMb78Its87Wn7flQ2B6GlU9A9zGg34XvG3JMGUZaVMzmFOH7PHS1SVWLMoz2bJXGwlsPtztI76GtU1eOPBXFHpnkH8fA1Iq1B5Ir0caIjtGmSYGcUyV0yrykkCHO0Mfrz1j6jZfJzj4T8IHht5INTfSCM2e-b28cBVPf-3eJutSKKDKOw5SfpABJiVjakxJs8jh6tPmuZPLaaFZ9Eu8Qn5PI7W-pZ1niC4Um2V_7EgI7ZKu2u0wY-Qcs5HuVEsPkJmwyIHo_SY1npcVZUKH5CxebcCxd6Z2XnjL9NLGnIqChD4gKukKZFdAL080viTk1S8F6Je_ar6yR9A9y7QHX9dSNo8xqusFTvxaqhBNr9iwyy8hNZljvoSfE6VH6bpQOZ_rGsMr1LQNfAiekNxHwe9fRTxPR33quwg_thDUsF3K9XflDI0Q36Nr2sVZqE0_0fDtuE3Cprlrd44iav7sjusbt8XAofVmycjq0zgpez7BZ5iBWTNSsiIrc0-0lIctyE04OV31vcjUhEBbPG5_aoCF_dgUSfCyifvJzDVYSISOPRjdyNbixLtS-7DKl0Wnb3hGmaflGeJhedC6xQH8gBNUl6S6Vh7Vo5WhMA1U7Q6nLXF1DWwY8ltScXk5m2VWDswsBRPx0oU3R6JZ-d4qwKARIA4f--WOoS_7P8SkRuTFQKgJ6ZbL6872Ch9dWswemjKMhug2uHw2TM_7i_r3xHgCe3wJik4ujHN8oSiZZALfQPymkV0fSfdaiVaqHP-aMQ_B0KBFlm7gF0D3XHbkWjcFIsLQ6NA49a-LDZWsG0UvPsQNGSbEDxyppveyzJJyOxescdgIGZ4A&isca=1) of a one of the random runners on this repo, it took about 15 seconds for the user data script to run.
> 
> My suspicious is on AWS, the time they take to hand over a VM is quite slow. Let me know if you have more questions. I am happy to jump on a call with you and AWS to see where the bottlenecks are.

It seems that after the runner is registered to become the idle state, it will switch to the offline state, and then it will become the Active state

**Comment by mytting** [comment]
I tried it. It took about 25 seconds from creating EC2 to being able to ssh. Is there a network reason for downloading the product?

**Comment by aktech** [comment]
> ok ，I have a general understanding, is there a general script? How to pass it to EC2, I can let AWS people debug it

I can try to create one for you.

> It seems that after the runner is registered to become the idle state, it will switch to the offline state, and then it will become the Active state

Ah, interesting.

> I tried it. It took about 25 seconds from creating EC2 to being able to ssh. Is there a network reason for downloading the product?


Did you create it via API? Can you share the script? If that's the case then it might be something on our end. I am happy to take a look, later this week.

**Comment by mytting** [comment]
Manually created... ..

**Comment by mytting** [comment]
I mean I created it manually and didn't pass in user-data?

**Comment by aktech** [comment]
> I mean I created it manually and didn't pass in user-data?

Ah, ok. I'll debug ours and will let you know.

**Comment by mytting** [comment]
ok

**Comment by mytting** [comment]
---
title: DeepFlow Accelerates GitHub Action Exploration Using Spot Instances
date: 2022/11/01
author: Song Jianchang
avatar:
cover: https://yunshan-guangzhou.oss-cn-beijing.aliyuncs.com/pub/pic/20221027635a6171c75b3.png
excerpt:
 - Github Action makes the CI process of projects hosted in Github very convenient, but the Runner configuration of 2C7G provided by Github by default is too low, and it will be very slow to run some large-scale project compilation tasks. In the exploration of Action, after a series of pitfalls, we finally found an ideal solution to solve all the requirements of performance, cost, ARM, etc. We hope it will be useful to you.
tags:
 - Github Action
 -Spot
 - Cirun
 - CI
---

Github Action makes the CI process of projects hosted in Github very convenient, but the Runner configuration of 2C7G provided by Github by default is too low to run some large-scale project compilation tasks. This article is an exploration of DeepFlow using public cloud high-end cheap Spot instances to accelerate Action , After a series of stepping on the pit, we finally found an ideal solution to solve all the needs of performance, cost, ARM, etc. I hope it will be useful to you.

# 0x0: Problems with GitHub Actions

Since the DeepFlow open source code was pushed to GitHub, we encountered the problem that GitHub Action took too long to compile tasks due to the low managed Runner configuration. Before that, the Alibaba Cloud 32C ECI Spot instance used by our internal GitLab CI took a few minutes. You can run all the jobs (the specific method will be introduced in a separate article later). After seeing the changes brought by the Spot instance to our GitLab CI, the GitHub Action of DeepFlow has been looking for a way from the first day of its launch. Each Job is assigned an independent Runner and supports the solution of the X86/ARM64 architecture, but this process is not smooth, and after 5 versions before and after, an ideal solution is finally found.

Some problems encountered by DeepFlow in the early stage of using GitHub Action:

1. Poor performance: The GitHub hosting machine configuration is too low and the compilation phase takes too long
2. Poor flexibility: The fixed `self-hosted` Runner cannot be dynamically scaled, the jobs are often queued, the machine is idle for a long time, and it is not cost-effective to fill the annual and monthly configuration
3. High cost: ARM64 architecture compilation requires another machine, Alibaba Cloud does not have ARM64 architecture instances in overseas regions, and AWS’s annual and monthly ARM64 machine price is relatively high (about 500$ per month for 32C64G)
4. Network instability: GitHub-hosted Runner often times out when pushing images to domestic Alibaba Cloud warehouses

Our needs:

- Support auto scaling, assign at least one Runner with at least 32 cores to each Job
- Works in much the same way as a GitHub-hosted Runner
- Supports both X86 and ARM64 architectures
- Low maintenance cost, preferably no maintenance
- Stable and fast push image to domestic Alibaba Cloud image warehouse

# 0x1: Accelerate the exploration of GitHub Actions

Based on our needs and the GitHub Action community documentation, we also found some solutions:
1. [K8s Controller](https://github.com/actions-runner-controller/actions-runner-controller): Kubernetes controller for GitHub Actions self-hosted runner
2. [Terraform](https://github.com/philips-labs/terraform-aws-github-runner): Autoscale AWS EC2 as GitHub Runner with Terraform and AWS Lambda
3. [Github](https://docs.github.com/en/actions/using-github-hosted-runners/using-larger-runners): Paid Larger Runners service currently only open to GitHub Team and Enterprise organizations
4. [Cirun](https://cirun.io): Automatically scale VMs of cloud platforms such as AWS/GCP/AZURE/OpenStack as GitHub Runner

| | **K8s Controller** | **Terraform** | **Github Larger Runners** | **Cirun** |
| :-------------- | :---------------------------- | :-- --------------------------- | :---------------------- --- | :------------------------- |
| Runner | Container | Linux, Windows | Linux, Window, Mac | Linux, Windows, Mac |
| Supported Cloud Platforms | Kubernetes | AWS | - | AWS, GCP, Azure, OpenStack |
| Whether to support ARM64 | Supported | Supported | Not supported | Supported |
| Whether Spot is supported | Not supported | Supported | - | Supported |
| Deployment Maintenance Cost | Medium | High | None | None |

The first solution we tried was `K8s Controller`. After trying it out, we found the following defects:

- AWS EKS Fargate does not support Privileged Containers
- AWS EKS Fargate does not support ARM64 architecture
- AWS EKS Fargate does not support `generic-ephemeral-volumes`

If there is no way to use Fargate, it is necessary to prepare independent nodes, and there is no way to dynamically scale nodes and use pay-as-you-go instances and Spot instances.

Next we tried the `Terraform` solution, but also encountered some setbacks:

- Failed to install and deploy multiple times
- Not friendly to Terrafrom newbies
- there is maintenance cost

The `GitHub` solution does not support ARM64 instances, and passes directly.

In the end we chose Cirun:

Cirun: Supports customization of arbitrary machine specifications, architectures and images. It is free for open source projects, does not require deployment and maintenance, does not require additional resources, and is very simple to operate:

**Step 1: Install the App**
  Install Cirun APP in [GitHub Marketplace](https://github.com/marketplace/cirun-io)

   ![Install Cirun](https://yunshan-guangzhou.oss-cn-beijing.aliyuncs.com/pub/pic/20221031635f9ad2024c3.png)

**Step 2: Add Repo**
  Add the required repo in the [Cirun console](https://cirun.io/admin/repositories)

   ![Add Repo](https://yunshan-guangzhou.oss-cn-beijing.aliyuncs.com/pub/pic/20221031635f9ad117d77.png)

**Step 3: Configure AK/SK**
  Configure AWS `ACCESS KEY` and `Secret KEY` in the [Cirun console](https://cirun.io/admin/repositories)

   ![AWS AUTH](https://yunshan-guangzhou.oss-cn-beijing.aliyuncs.com/pub/pic/20221031635f9ad218de9.png)

**Step 4: Configure Machine Specifications**
Machine specification and Runner Label are defined in GitHub Repo
````yaml
runners:
  - name: "aws-amd64-32c"
    cloud: "aws"
    instance_type: "c6id.8xlarge"
    machine_image: "ami-097a2df4ac947655f"
    preemptible: true
    labels:
      - "aws-amd64-32c"
  - name: "aws-arm64-32c"
    cloud: "aws"
    instance_type: "c6g.8xlarge"
    machine_image: "ami-0a9790c5a531163ee"
    preemptible: true
    labels:
      - "aws-arm64-32c"
````
**Step 5: Getting Started**
Toggle GitHub Job's `runs-on` field
````yaml
jobs:
  build_agent:
    name: build agent
    runs-on: "cirun-aws-amd64-32c--${{ github.run_id }}"
    steps:
      - name: Checkout
        uses: actions/checkout@v3
        with:
          submodules: recursive
          fetch-depth: 0
````

**final effect**:

![GitHub Runner](https://yunshan-guangzhou.oss-cn-beijing.aliyuncs.com/pub/pic/20221031635f94aee6ccf.jpg)
![AWS Instance](https://yunshan-guangzhou.oss-cn-beijing.aliyuncs.com/pub/pic/20221031635f94ad43815.jpg)

# 0x2: Changes after DeepFlow uses Cirun

Currently DeepFlow uses AWS 32C64G Spot instances to run CI in parallel, with an average monthly consumption of $300. If you use the monthly subscription method, you can only run two 16C32G X86/ARM64 instances under the same consumption, and once the

----------------
@aktech 

This is a recent article I wrote to promote cirun. Can you give me some advice? Can you also give a brief introduction to what cirun is？

**Comment by mytting** [comment]
thanks

**Comment by mytting** [comment]
> 是的，这是一个突出的错误。它将在下一个版本中修复。

@aktech Hello, is there any progress on this issue? In recent days, X86 machines also have spot requests, which often affects CI.

**Comment by aktech** [comment]
By the way is that blog post already published, the one you mentioned above? If yes, can you share the link please?

**Comment by mytting** [comment]
Yes, but only in Chinese. Is it convenient for you to check and share?
https://mp.weixin.qq.com/s/26qbfq7bBmmgOk_NFWVUow
https://deepflow.yunshan.net/blog/015-deepflow-uses-spot-Instances-to-speed-up-github-action-exploration/

**Comment by aktech** [comment]
Awesome, thanks a lot! i just used google translate to view the page. Is it possible for you to post it in english as well (just the google translate) on somewhere like say [dev.to](https://dev.to/)? I am happy to review the translation.

**Comment by mytting** [comment]
I'll try it next week.

**Comment by aktech** [comment]
Sure, no hurries.

**Comment by aktech** [comment]
I wrote a twitter thread, quoting from your blog: https://twitter.com/iaktech/status/1593574852241154049

**Comment by aktech** [comment]
@mytting I have changed the backend to use the latest `CreateFleet` API of AWS for creating spot instances:
https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/spot-best-practices.html#which-spot-request-method-to-use

I did not see much difference in the runner creation time for smaller instances (like `t2.medium`), let me know if you see any difference (for your larger instances). I have also done some profiling to see if there are any bottlenecks anywhere else, I haven't found any so far apart from runner creation from AWS itself.

**Comment by mytting** [comment]
@aktech Is it a launch parameter passed through user-data? Is it convenient to provide a general user-data script? Let me test the difference between adding user-data and not.

**Comment by mytting** [comment]
> | Syntax      | Description |
> | ----------- | ----------- |
> | Header      | Title       |
> | Paragraph   | Text        |

https://dev.to/dundun/deepflow-uses-spot-instances-to-speed-up-github-action-exploration-2a90

**Comment by mytting** [comment]
> > @aktech Is it a launch parameter passed through user-data? Is it convenient to provide a general user-data script? Let me test the difference between adding user-data and not.
> 
> So, the launch template is created first (with user-data) via: [create_launch_template](https://boto3.amazonaws.com/v1/documentation/api/latest/referenc/services/ec2.html#EC2.Client.create_launch_template) and then it is passed to [create_fleet](https://boto3.amazonaws.com/v1/documentation/api/latest/reference/services/ec2.html#EC2.Client.create_fleet).
> 
> > https://dev.to/dundun/deepflow-uses-spot-instances-to-speed-up-github-action-exploration-2a90
> 
> Excellent, thanks a lot!


OK, I'm interested in what is the general content of the user-data? The vm I created in the console started up quickly.

**Comment by aktech** [comment]
> OK, I'm interested in what is the general content of the user-data? The vm I created in the console started up quickly.

Sure, I can share that. Can you join slack [here](http://slack.cirun.io/). I'll DM you, after I create one minimal example.

**Comment by mytting** [comment]
ok
